waspc/waspls/test/Wasp/LSP/CompletionTest.hs

```diff
@@ -0,0 +1,179 @@
+--
+-- See 'CompletionTestInput' for the format of the test input.
+parseCompletionInput :: CompletionTestInput -> (String, LSP.Position)
+parseCompletionInput (CompletionTestInput testInput) =
+  withPrefixAssert "//! test/completion" testInput (waspSourceCode, cursorPosition)
+  where
+    -- 1. Get the list of consecutive lines in @testInput@, with the line number
+    -- of the first line.
+    linePairs = zip3 (lines testInput) (drop 1 $ lines testInput) [0 ..]
+
+    -- 2. Get list of (String, column, line) triples representing each line
+    -- that is followed by a "carrot line" with the carrot at the specified
+    -- column.
+    --
+    -- See 'CompletionTestInput' for a description of a "carrot line".
+    candidateLines = mapMaybe (makeCandidateLine 0) linePairs
+
+    -- 3. Find the first candidate line with a "|" at the correct column,
+    -- indicating that this is where the cursor is located in the input.
+    (lineWithCursor, cursorColumn, cursorLine) =
+      case find isLineWithCursor candidateLines of
+        Nothing -> error "readCompletionTest: no marked line"
+        Just x -> x
+
+    -- 4. Remove the "|" from the line that contained the cursor.
+    lineWithCursorCleanedUp = before cursorColumn lineWithCursor ++ after cursorColumn lineWithCursor
+
+    -- 5. Write the wasp source code represented by the test input by removing
+    -- the "|" from the line with the cursor and removing the entire "carrot
+    -- line."
+    waspSourceCode =
+      unlines $
+        concat
+          [ before cursorLine (lines testInput),
+            [lineWithCursorCleanedUp],
+            -- cursorLine + 1 so the line with the ^ is also removed.
+            after (cursorLine + 1) (lines testInput)
+          ]
+
+    cursorPosition = LSP.Position (fromIntegral cursorLine) (fromIntegral cursorColumn)
+
+    -- Check if the (String, column, line) contains a '|' at the specified column.
+    isLineWithCursor :: (String, Int, Int) -> Bool
+    isLineWithCursor (str, column, _line) = (length str >= column) && ((str !! column) == '|')
+
+    -- @makeCandidate line (firstLine, secondLine)@ checks that the second line
+    -- is a "carrot line" and returns a triple  @(firstLine, column, line)@,
+    -- where @column@ is the column that the carrot is at on the second line.
+    makeCandidateLine :: Int -> (String, String, Int) -> Maybe (String, Int, Int)
+    makeCandidateLine n (a, ['^'], ln) = Just (a, n, ln)
+    makeCandidateLine n (a, ' ' : bs, ln) = makeCandidateLine (n + 1) (a, bs, ln)
+    makeCandidateLine _ _ = Nothing
+
+    -- Check that the string begins with a prefix, throwing an error if not.
+    withPrefixAssert :: String -> String -> a -> a
+    withPrefixAssert preamble str x
+      | (preamble ++ "\n") `isPrefixOf` str = x
+      | otherwise = error $ "test expected to begin with preamble: " ++ preamble
+
+    before :: Int -> [a] -> [a]
+    before = take
+
+    after :: Int -> [a] -> [a]
+    after idx = drop (idx + 1)
```

Comment:
I found this to be a bit imperatively written / hard to read, so I ended up giving it a go of my own:

```hs
-- | Parses a completion test case into a pair of the wasp source code to
-- run completion on and the position to get the completion list at.
--
-- See 'CompletionTestInput' for the format of the test input.
parseCompletionInput :: CompletionTestInput -> (String, LSP.Position)
parseCompletionInput (CompletionTestInput waspSourceWithCursor) =
  if not (hasCompletionTestPreamble waspSourceWithCursor)
    then error missingCompletionTestPreambleErrorMsg
    else (waspSource, cursorPosition)
  where
    waspSource =
      unlines . concat $
        [ linesBeforeCursor,
          [cursorLineWithCursorRemoved],
          linesAfterCursor
        ]

    cursorPosition = LSP.Position (fromIntegral cursorLineIdx) (fromIntegral cursorColumnIdx)

    linesBeforeCursor = take cursorLineIdx (lines waspSourceWithCursor)

    -- NOTE: We have +2 here to ensure we skip the "      ^   " line after the cursor.
    linesAfterCursor = drop (cursorLineIdx + 2) (lines waspSourceWithCursor)

    cursorLineWithCursorRemoved =
      let cursorLine = lines waspSourceWithCursor !! cursorLineIdx
       in take cursorColumnIdx cursorLine <> drop (cursorColumnIdx + 1) cursorLine

    (cursorLineIdx, cursorColumnIdx) =
      fromMaybe (error "parseCompletionTest: found no cursor!") . msum $
        (\(lineIdx, linePair) -> (lineIdx,) <$> cursorColIdxInConsecutiveLinesPair linePair)
          <$> zip [0 ..] (zip (lines waspSourceWithCursor) (drop 1 $ lines waspSourceWithCursor))

    -- Given two consecutive lines from the test input, find cursor if there is one
    -- and return the index of the column it is in.
    cursorColIdxInConsecutiveLinesPair :: (String, String) -> Maybe Int
    cursorColIdxInConsecutiveLinesPair (line1, line2) = do
      cursorColIdx <- elemIndex '|' line1
      if (length line2 > cursorColIdx)
        && all (\(idx, c) -> if idx == cursorColIdx then c == '^' else isSpace c) (zip [0 ..] line2)
        then return cursorColIdx
        else Nothing

    hasCompletionTestPreamble :: String -> Bool
    hasCompletionTestPreamble source = (completionTestPreamble <> "\n") `isPrefixOf` source

    missingCompletionTestPreambleErrorMsg =
      "test expected to begin with preamble: " <> completionTestPreamble

    completionTestPreamble = "//! test/completion"
```

I removed `withPrefixAssert` because I found it too abstract, hard to understand when I see it up there what it does, so I inlined it instead.
As for the rest of the code, I organized it more declaratively, so more top down versus steps from the bottom to the top.

Take a look if you like it and use it if so (or parts of it)!

Btw I am starting to think this could have been made a bit easier if we used something like `{CURSOR}` instead of `|` and then `^` in the next line :D. It does look visually appealing, but it does also force that blank line and is harder to parse. Ok, it also has much less chance (0?) of occuring naturaly in Wasp code, but still, we control this code so in practice there will not happen that we have {CURSOR} somewhere in it.
But ok, we already have this and it is kind of nice so if you want to keep the | + ^ I think that is ok, it is cool at the end.

Reply:
> As for the rest of the code, I organized it more declaratively, so more top down versus steps from the bottom to the top.

Ah, that's funny! I just reorganized it to be more imperative because it felt like it might be more readable (if not less Haskellish). But I'll switch it back to be declarative again.

I've pulled a bunch of the ideas from your suggested code too.

Reply:
Ah sorry about that! Sometimes it can be more readable if imperative, but I do like when declarative code goes from top to bottom, I think that is one of its main attractive attributes, especially when order of actions is not important.